### DISCUSSION TITLE
Sorting a Grid

### DISCUSSION THREAD

**Discussion by rossaddison** [yiisoft/docs#220] [Show and tell]
![image](https://github.com/user-attachments/assets/92fd407c-41c3-4f9a-86d0-38d894d100d1)
![image](https://github.com/user-attachments/assets/1d5f876b-a9a9-4fb9-ad13-ff0f119764b8)

Code: 

````
echo GridView::widget()
    // unpack the contents within the array using the three dot splat operator
    ->rowAttributes(['class' => 'align-left'])
    ->tableAttributes(['class' => 'table table-striped h-75', 'id' => 'table-invoice'])
    ->columns(...$columns)
    ->dataReader($paginator)
    ->headerTableEnabled(true)        
    ->headerRowAttributes(['class' => 'card-header bg-info text-black'])
    ->footerEnabled(true)    
    ->footerRowAttributes(['id' => 'this is for id'])
    ->emptyCell($translator->translate('i.not_set'))
    ->emptyCellAttributes(['style' => 'color:red'])        
    ->sortableHeaderPrepend('<div class="float-end text-secondary text-opacity-50">⭥</div>')        
    ->sortableHeaderDescAppend(A::tag()
        ->content('⬆')
        ->href($urlGenerator->generate('inv/index', [], ['sort' => 'id']))
        ->id('btn-id-desc-append')
        ->render())
    ->sortableHeaderDescPrepend(A::tag()
        ->content('⬇')
        ->href($urlGenerator->generate('inv/index', [], ['sort' => '-id']))
        ->id('btn-id-desc-prepend')
        ->render())
    ->sortableHeaderAscAppend(A::tag()
        ->content('⬆')
        ->href($urlGenerator->generate('inv/index', [], ['sort' => 'id']))
        ->render())                
    ->sortableHeaderAscPrepend(A::tag()
        ->content('⬇')
        ->href($urlGenerator->generate('inv/index', [], ['sort' => '-id']))
        ->id('btn-id-asc-prepend')
        ->render())        
    ->header($gridComponents->header(' ' . $translator->translate('i.invoice')))
    ->id('w3-grid') 
    ->pagination(
        $gridComponents->offsetPaginationWidget($defaultPageSizeOffsetPaginator, $paginator)
    )
    ->summaryAttributes(['class' => 'mt-3 me-3 summary text-end'])    
    ->summaryTemplate($pageSizeLimiter::buttons($currentRoute, $s, $urlGenerator, 'inv').' '.$grid_summary)
    ->emptyTextAttributes(['class' => 'card-header bg-warning text-black'])
    ->emptyText($translator->translate('invoice.invoice.no.records'))
    ->toolbar($toolbarString);     
````
How can the above code be improved?
Weakness: It only allows for one sorting column namely the 'id'.  I want to move the 
![image](https://github.com/user-attachments/assets/253fd2a8-fa38-4439-b319-19d2b71c9fad)
![image](https://github.com/user-attachments/assets/dfaac98c-f2ad-4892-839c-6cb6ad5dd394)
buttons into their own header and ideally there should be some configuration file/array that I can adjust.

**Comment by vjik** [comment]
`GridView` gets sort information from data reader. Set `Sort` object to your data reader. For example:

```php
$sort = Sort::only([
	'status_id',
	'phone_id',
	'sales_manager_id',
	'creation_date',
	'modified_date',
])->withOrder(['creation_date' => 'desc']);
$dataReader = $dataReader->withSort($sort);
```

**Comment by rossaddison** [comment]
InvController.php

````
$sortString = $querySort ?? '-id';
            $order =  OrderHelper::stringToArray($sortString);
            $sort = Sort::only(['id', 'status_id', 'number', 'date_created', 'date_due', 'client_id'])
                    // (@see vendor\yiisoft\data\src\Reader\Sort
                    // - => 'desc'  so -id => default descending on id
                    // Show the latest quotes first => -id
                    ->withOrder($order);
````


How do you get the ascending and descending arrows to appear in each header? I can see if you remove 

````
new DataColumn(
            'id',
            header:     <--------------------------------------------------- remove
            content: static fn(Inv $model) => $model->getId(),
            withSorting: true
        ), 
````

the header from the DataColumn you then get a hyperlinked id appearing . i.e.
![image](https://github.com/user-attachments/assets/0d5b9524-08bb-420b-af1d-f01232c0393e)

But I want to be able to click on the header arrows enclosed in a query string i.e. something like
````
(A::tag()
        ->content('⬆')
        ->href($urlGenerator->generate('inv/index', [], ['sort' => 'id']))
        ->id('btn-id-desc-append')
        ->render()
````
...with the query parameter. In the above case the 'id' is not '-id' because it is ascending. I can get full functionality with the two id buttons outside the grid but I want to incorporate them inside the grid header above the id column.

If a incorporate the A::tag() in the header I get this effect: 

![image](https://github.com/user-attachments/assets/cf2c5bb2-da47-4cb6-a3bf-1887b3eae226).

**Comment by vjik** [comment]
Currently, header format is `%SortPrepend% %Label% %SortAppend%`.

If avaliable sort by  column, then `%Label%` will be link, that change sort. 

To generate URL grid view use url creator that you can set via method `urlCreator()`. Try to use `Yiisoft\Yii\DataView\YiiRouter\UrlCreator`.

Also you can set options:

- `sortableHeaderClass`
- `sortableHeaderPrepend`
- `sortableHeaderAppend`
- `sortableHeaderAscClass`
- `sortableHeaderAscPrepend`
- `sortableHeaderAscAppend`
- `sortableHeaderDescClass`
- `sortableHeaderDescPrepend`
- `sortableHeaderDescAppend`
- `sortableLinkAttributes`
- `sortableLinkAscClass`
- `sortableLinkDescClass`

**Comment by rossaddison** [comment]
Ok thanks Vjik. I have moved the ````status_id````, and the ````date due```` into the grid. I left the ````client_id```` in as a button. There appears to be a conflict when sorting and filtering are available on the column simultaneously.  

Additional code: InvController.php
````
            $urlCreator = new UrlCreator($this->url_generator);
            $order =  OrderHelper::stringToArray($sortString);
            $urlCreator->__invoke([], $order);
            $sort = Sort::only(['id', 'status_id', 'number', 'date_created', 'date_due', 'client_id'])
                    // (@see vendor\yiisoft\data\src\Reader\Sort
                    // - => 'desc'  so -id => default descending on id
                    // Show the latest quotes first => -id
                    ->withOrder($order);
            $invs = $this->invs_status_with_sort($invRepo, $status, $sort);

             The  $urlCreater appears in the parameters to the index as well.
````

Additional code: inv/index
````
->urlCreator($urlCreator)
    // the up and down symbol will appear at first indicating that the column can be sorted 
    // It also appears in this state if another column has been sorted        
    ->sortableHeaderPrepend('<div class="float-end text-secondary text-opacity-50">⭥</div>')
    
    // the up arrow will appear if column values are ascending in value         
    ->sortableHeaderAscPrepend('<div class="float-end fw-bold">⭡</div>')
   
    // the down arrow will appear if column values are descending in value       
    ->sortableHeaderDescPrepend('<div class="float-end fw-bold">⭣</div>') 


      
````
Full code:
[
inv/index.php](https://github.com/rossaddison/invoice/blob/main/resources/views/invoice/inv/index.php)

[InvController function index](https://github.com/rossaddison/invoice/blob/main/src/Invoice/Inv/InvController.php)

![image](https://github.com/user-attachments/assets/dd0758be-330e-43c8-b011-daee667fbd21)

**Reply by vjik** [reply]
It should work well with filter and sort at the same time. What error is appearing?

**Reply by rossaddison** [reply]
Ok here is my DataColumn: 
````
new DataColumn(
            field: 'client_id',
            property: 'filterClient',
            header: $translator->translate('i.client'),
            content: static fn(Inv $model): string => (string)$model->getClient()?->getClient_full_name(),
            filter: $optionsDataClientsDropdownFilter,
            withSorting: true
        ),
````
![image](https://github.com/user-attachments/assets/367b3715-f1b9-4f04-b59c-3ac883edc1b0)

It sorts on first click but on second click it does not reverse the sort order and 'filterClient' appears in the query parameter on the first click and on the second click.

**Reply by rossaddison** [reply]
[I have put in a pull request. Hopefully might help. ](https://github.com/yiisoft/yii-dataview/pull/216/files) . The originalProperty  value e.g. 'filterClient', is replacing the property value e.g 'client_id' by means of the code: 

function getLinkSortValue
````
return OrderHelper::arrayToString(
            ArrayHelper::renameKey($resultOrder, $property, $originalProperty)
        );
````

which is why the sort link is changing to 'filterClient' which is incorrect. The fix is working fine.

**Reply by vjik** [reply]
Field property is used in `sort` parameter of URL. It replaced to `field` before pass it to data reader. We don't use `field` in URL because it's not safe. Field may contains sensible data, for example, table name: `my_db_table.field_name`.

`GridView` process `sort` parameter of URL itself. You not need process it in your controller.

All filters and sort should be processed by `GridView`, not need extra code for it.

For example, [here](https://github.com/rossaddison/invoice/blob/f2932096634db43f7f55f26fdccad93f2780f4d2/src/Invoice/Inv/InvController.php#L2009) you should set default order. `GridView` itself change order, if it passed by URL.

**Reply by rossaddison** [reply]
Ok Vjik, I have pulled my sorting code from the Controllers. I still have some filtering code in the Controllers. Do you perhaps have a working example of a grid with the code somewhere? Would be really useful on the demo.
